packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -3,69 +3,69 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:ui';
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/painting.dart';
 
-/// A leaf node in the focus tree that can receive focus.
-///
-/// The focus tree keeps track of which widget is the user's current focus. The
-/// focused widget often listens for keyboard events.
-///
-/// To request focus, find the [FocusScopeNode] for the current [BuildContext]
-/// and call the [FocusScopeNode.requestFocus] method:
-///
-/// ```dart
-/// FocusScope.of(context).requestFocus(focusNode);
-/// ```
-///
-/// If your widget requests focus, be sure to call
-/// `FocusScope.of(context).reparentIfNeeded(focusNode);` in your `build`
-/// method to reparent your [FocusNode] if your widget moves from one
-/// location in the tree to another.
-///
-/// ## Lifetime
+import 'binding.dart';
+import 'focus_traversal.dart';
+import 'framework.dart';
+
+typedef FocusNodeVisitor = bool Function(FocusNode node);
+
+/// A node used by [Focus] to represent and change the input focus.
 ///
-/// Focus nodes are long-lived objects. For example, if a stateful widget has a
-/// focusable child widget, it should create a [FocusNode] in the
-/// [State.initState] method, and [dispose] it in the [State.dispose] method,
-/// providing the same [FocusNode] to the focusable child each time the
-/// [State.build] method is run. In particular, creating a [FocusNode] each time
-/// [State.build] is invoked will cause the focus to be lost each time the
-/// widget is built.
+/// A [FocusNode] is created and managed by the [Focus] widget, forming
+/// a tree of nodes that represent things that can be given the input focus in
+/// the UI.
 ///
-/// See also:
+/// [FocusNode]s are persistent objects which are reparented and reorganized
+/// when the widget tree is rebuilt.
 ///
-///  * [FocusScopeNode], which is an interior node in the focus tree.
-///  * [FocusScope.of], which provides the [FocusScopeNode] for a given
-///    [BuildContext].
-class FocusNode extends ChangeNotifier {
-  FocusScopeNode _parent;
+/// The can be organized into "scopes", which form groups of nodes which are to
+/// be traversed together. Within a scope, the last item to have focus is
+/// remembered, and if another scope is focused and then the first is returned
+/// to, that node will gain focus again.
+class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
```

Comment:
per our discussion, would be good to put a book about focus here

Reply:
Yeah, already well on my way.